### PR TITLE
[6716] Put base64 methods in irods namespace

### PR DESCRIPTION
**Pull request by JustinKyleJames**

### CODE REVIEW

File: lib/core/include/irods/base64.h
Comment (alanking) [inline lib/core/include/irods/base64.h:20]: Since we are removing the C linkage, should we rename this file to base64.hpp to further indicate that C clients should not be using it?

**Review by alanking — COMMENTED** [review]
A lot of the diff seems to be a clang-format run, so maybe run that once more to make sure it's happy.

Had another question/concern in the header file as well.

Also-also, more words in the commit message could be helpful. I know the issue will be linked, but, y'know. ;)

File: lib/core/include/irods/base64.h
Comment (korydraughn) [inline lib/core/include/irods/base64.h:12]: This doesn't look like it is needed.

Try removing it and seeing if things still compile. If it fails to compile, put this `#include` in the cpp file.

File: lib/core/src/base64.cpp
Comment (korydraughn) [inline lib/core/src/base64.cpp:48]: Let's move this to the header file.

The contents needs to be fixed to match how we document things now. See https://github.com/irods/irods/blob/c3c62ad4cfe8064bcc1475839cc79f362d0e1caa/lib/administration/user/include/irods/user_administration.hpp for examples.

File: lib/core/src/base64.cpp
Comment (korydraughn) [inline lib/core/src/base64.cpp:94]: Let's move this to the header file.

File: lib/core/src/base64.cpp
Comment (korydraughn) [inline lib/core/src/base64.cpp:8]: Consider removing this and replacing all usage with `0`.

File: lib/core/src/base64.cpp
Comment (korydraughn) [inline lib/core/src/base64.cpp:6]: Please swap the order of these includes and delete all lines above them.

**Review by korydraughn — COMMENTED** [review]
The code changes are good.

My comments are mostly about modernizing the structure of the file. For these smaller files, we should go ahead and improve them when we touch them. That may not be possible for all instances, but these base64 files seem perfectly safe to update.

I can't leave comments on some of the things, so I'll mention them here.
- Delete everything before `#ifndef BASE64_H__`
- Change `BASE64_H__` to `IRODS_BASE64_H`
- Insert `/// \file` before the `#include`s in the header file

You can do that in a separate commit and attach issue number 858 to it.

File: lib/core/include/irods/base64.h
Comment (korydraughn) [reply lib/core/include/irods/base64.h:20]: Given that this change breaks anyone originally relying on these functions, it is okay to change the file extension.

After saying that, we need to think about this a bit more.

Given that the dev pkg exposes almost everything used in iRODS, we can't break people's code. We need to provide a solution that let's the user transition gracefully or we need to define a list of things people are allowed to use.

Below is a potential plan for the transitional approach.
- Leave the old functions in base64.h and mark them deprecated
- Create a new file with the .hpp extension and copy @JustinKyleJames's new version of base64.h into it
- Add a `#warning` to the base64.h file that let's devs know the header is deprecated and that they shouldn't use it anymore. The warning could point the dev to the cpp file or maybe we don't ship the hpp version at all

Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h:20]: Are clients using this directly?  I have my doubts whether they are.

The Globus client only used it indirectly via the clientLogin() call.

**Comment by JustinKyleJames** [comment]
This passed all testing.  I will run again once I make the changes mentioned above.

File: lib/core/include/irods/base64.h
Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h:12]: done

File: lib/core/src/base64.cpp
Comment (JustinKyleJames) [reply lib/core/src/base64.cpp:6]: done

File: lib/core/include/irods/base64.h
Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h:12]: done

File: lib/core/src/base64.cpp
Comment (JustinKyleJames) [reply lib/core/src/base64.cpp:48]: done

File: lib/core/src/base64.cpp
Comment (JustinKyleJames) [reply lib/core/src/base64.cpp:94]: done

File: lib/core/src/base64.cpp
Comment (JustinKyleJames) [reply lib/core/src/base64.cpp:8]: done

File: lib/core/src/base64.cpp
Comment (JustinKyleJames) [reply lib/core/src/base64.cpp:6]: done

File: lib/core/src/base64.cpp
Comment (JustinKyleJames) [reply lib/core/src/base64.cpp:94]: done

**Review by korydraughn — COMMENTED** [review]
Looks good to me.

The only remaining question is whether we should instruct people to use the base64.hpp header (as mentioned in the warning message).

I think it is fine overall. The functions are in the irods namespace now and their implementations are correct. This means these functions are only available to C++ developers.

Making base64.hpp private can be handled in a separate PR.

File: lib/core/include/irods/base64.h
Comment (alanking) [reply lib/core/include/irods/base64.h:6]: I still see the `#include`...

File: lib/core/src/base64.cpp
Comment (alanking) [inline lib/core/src/base64.cpp:112]: Should we mark the declarations in the header file with `__attribute__(deprecated)` as well?

File: lib/core/include/irods/base64.h
Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h]: The requested changes have been implemented.

File: lib/core/include/irods/base64.h
Comment (alanking) [inline lib/core/include/irods/base64.h:2]: I believe @korydraughn mentioned renaming these header guards to `IRODS_BASE64_H`

**Review by alanking — COMMENTED** [review]
Please ignore my comments if they're addressed in a to-be-pushed commit :)

File: lib/core/include/irods/base64.h
Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h:2]: He mentioned it for the HPP one.  I was intending to leave he old deprecated one alone although I did remove the first few lines.

Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h:2]: I suppose since I have already touched it I might as well rename the header guard.

Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h]: done

File: lib/core/src/base64.cpp
Comment (JustinKyleJames) [reply lib/core/src/base64.cpp:112]: I did this.  Please take a look to make sure it is done correctly.  I have never used that.

**Comment by JustinKyleJames** [comment]
I forgot to add/commit base64.hpp.  It has been added now.

File: lib/core/include/irods/base64.h
Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h:18]: From my testing it seems to work fine using these in a test program.  You just get the warning on compile.  After removing the warning on the header itself, iRODS compiles fine as it isn't actually calling these methods.

Comment (SwooshyCueb) [reply lib/core/include/irods/base64.h:18]: As we discovered previously, there's overlap with declarations from another library. I think it would be preferable to deprecate the header itself rather than the declarations, but I don't feel super strongly about it.

File: lib/core/src/packStruct.cpp
Comment (alanking) [inline lib/core/src/packStruct.cpp:2033]: Well, I quite hate this. :p

**Review by alanking — DISMISSED** [review]
Other than clang-format thing, seems good to me, assuming everything is building and tests pass.

File: lib/core/src/packStruct.cpp
Comment (korydraughn) [reply lib/core/src/packStruct.cpp:2033]: Yeah. It's not pretty. PackStruct code isn't pretty either lol.

I'm fine with letting this ride though.

File: lib/core/include/irods/base64.hpp
Comment (korydraughn) [inline lib/core/include/irods/base64.hpp:21]: Let's add a newline between these lines so that doxygen doesn't get tripped up by it and so that it matches the documentation on the function above.

**Comment by alanking** [comment]
I think everything has been addressed here? Please resolve comments if true and we shall await one more approval.

**Review by korydraughn — APPROVED** [review]
Looks good. Please squash it and leave off the pounds.

**Comment by JustinKyleJames** [comment]
Squashed everything into one pull request.

**Review by korydraughn — DISMISSED** [review]
There's one last comment that needs to be marked as resolved if it has been handled.

File: lib/core/include/irods/base64.h
Comment (JustinKyleJames) [reply lib/core/include/irods/base64.h]: The header has been deprecated and the declarations left alone.

**Comment by JustinKyleJames** [comment]
> There's one last comment that needs to be marked as resolved if it has been handled.

Done

**Comment by korydraughn** [comment]
Let's add the pound and get this merged.

**Review by SwooshyCueb — APPROVED** [review]

**Comment by JustinKyleJames** [comment]
Pound added to commit.

**Review by alanking — APPROVED** [review]

**Pull request merged**
